tests/test_indexes.py

```diff
@@ -1,58 +1,63 @@
 
 from . import PandasTestCase
 import unittest
 import string
 from parameterized import parameterized
+import importlib
+import inspect
+import numpy as np
 
 
-# Define valid inputs for different functions.
+"""
+This file intends to test each function whether the input Series's index is the same as
+the output Series.
+
+This will go through all functions under texthero, and automatically 
+generate test cases, according to the HeroSeries type they accept, e.g. TokenSeries, TextSeries, etc.
+
+Normally, functions receives HeroSeries and returns Series is okay for auto-testing.
+However there might be exceptions that you want to specify a test case manually, or omit a function
+for testing. For example, 
+- Functions that needs multiple arguments (preprocessing.replace_stopwords)
+- Functions that returns Series with different index (representation.tfidf)
+- Functions that doesn't give Series as output (mostly in visualization)
+- Functions that doesn't take HeroSeries (yet)
+
+In those cases, you can add your custom test case so as to override the default one, 
+in the form of [name_of_test_case, function_to_test, tuple_of_valid_input_for_the_function]. 
+If you want to omit some functions, add their string name to func_white_list variable.
+
+The tests will be run by AbstractIndexTest below through the @parameterized
+decorator. The names will be expanded automatically, so e.g. "named_entities"
+creates test cases test_correct_index_named_entities and test_incorrect_index_named_entities.
+"""
+
+# Define the valid input for each HeroSeries type
 s_text = pd.Series(["Test"], index=[5])
 s_tokenized_lists = pd.Series([["Test", "Test2"], ["Test3"]], index=[5, 6])
 s_numeric = pd.Series([5.0], index=[5])
 s_numeric_lists = pd.Series([[5.0, 5.0], [6.0, 6.0]], index=[5, 6])
 
-# Define all test cases. Every test case is a list
-# of [name of test case, function to test, tuple of valid input for the function].
-# First argument of valid input has to be the Pandas Series where we
-# want to keep the index. If this is different for a function, a separate
-# test case has to implemented in the class below.
-# The tests will be run by AbstractIndexTest below through the @parameterized
-# decorator.
-# The names will be expanded automatically, so e.g. "named_entities"
-# creates test cases test_correct_index_named_entities and test_incorrect_index_named_entities.
-
-test_cases_nlp = [
-    ["named_entities", nlp.named_entities, (s_text,)],
-    ["noun_chunks", nlp.noun_chunks, (s_text,)],
-]
+valid_inputs = {
+    "TokenSeries": s_tokenized_lists,
+    "TextSeries": s_text,
+    "VectorSeries": s_numeric_lists,
+}
+
+# Specify your custom test cases here (functions that
+# has multiple arguments, doesn't accpet HeroSeries, etc.)
+test_cases_nlp = []
 
 test_cases_preprocessing = [
-    ["fillna", preprocessing.fillna, (s_text,)],
-    ["lowercase", preprocessing.lowercase, (s_text,)],
     ["replace_digits", preprocessing.replace_digits, (s_text, "")],
```

Comment:
Can we generalize over these functions as well and avoid to define the custom test manually? Rewriting the file as you are proposing add more complexity for the developer, I can see it as worth it in case the custom functions that need to be manually defined are kept to the very minimum

Reply:
Can I assume that any functions with default params can be auto-tested?

Reply:
Seems reasonable, opinions?

Reply:
Then I think the simplest way might be regulating the function definition and add default params whenever possible, so the default params can be taken as a default test case.